### PR TITLE
draft

### PR DESCRIPTION
**Pull request by guiyom-e**

### CODE REVIEW

File: blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+cover_image: ''
+description: '6 main points to watch out when migrating data in databases.'
+tags: database, dynamodb
+series:
+canonical_url:
+---
+
+> *“The majority of project issues I have seen come from databases, whatever is the technology”*
+
+This is why I have been built 6 guidelines I try to follow when modifying data on tech projects. They are especially helpful when working with databases like DynamoDB with few structuring tools.
```
Comment (ealain) [inline blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md]:
```suggestion
This is why I have built 6 guidelines I try to follow when modifying data on tech projects. They are especially helpful when working with databases like DynamoDB with few structuring tools.
```

Comment (ealain) [reply blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md]: Tu peux préciser "Few structuring tools" ? Ça ne me semble pas être standard.

File: blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+published: false
+title: '6 guidelines for risk-less data migrations'
+cover_image: ''
+description: '6 main points to watch out when migrating data in databases.'
+tags: database, dynamodb
+series:
+canonical_url:
+---
+
+> *“The majority of project issues I have seen come from databases, whatever is the technology”*
```
Comment (ealain) [inline blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md]: Cool :) C'est de qui cette citation ?

File: blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+This is why I have been built 6 guidelines I try to follow when modifying data on tech projects. They are especially helpful when working with databases like DynamoDB with few structuring tools.
+
+> **Data is key in business… and it is where it often fails.**
+
+Almost every app has a **database** with more or less structured data. These data items may evolve during app development, with the addition of new fields, changes to field structure, data updates, or removal of fields. These evolutions, often referred to as **migrations**, are critical as they enable the app to evolve but can also introduce bugs if not executed properly.
+A data migration **plan** is a perfect tool to mitigate the risks of these operations, as it avoids common pitfalls and ensure reliability.
```
Comment (ealain) [inline blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md]:
```suggestion
A data migration **plan** is a perfect tool to mitigate the risks of these operations, as it avoids common pitfalls and ensures reliability.
```

File: blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+## 📋️ First thing: have a migration protocol defined
+
+### Define a protocol to follow in advance
+
+- What is the **RACI** ? (**Who** is responsible, who is informed, etc. ?)
```
Comment (ealain) [inline blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md]: Add a link to reference

File: blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+  - risk of maintaining multiple versions over a long period of time.
+
+## 🔂 Make a plan reproducible in multiple environments
+
+### Write idempotent migrations (i.e. a migration can be applied multiple times with the same result)
+
+- There are many reasons for needing to re-apply a migration: migration partially completed, stopped or failed, etc.
+- The migration script should work without any state-full input
+
+✍️ Simple example of **idempotent migrations**
```
Comment (ealain) [inline blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md:85]: Cool

File: blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+
+- Committing migration scripts ensures the same process can be applied to another environment easily and with known and reproducible steps.
+- In case of error, it will help understanding what happened.
+
+### Test your scenario
+
+- Test the migration script (with unit tests) to verify its functionality.
+  - Check that data items of the migration set are correctly selected
+  - Check all edge cases
+  - Try to keep the migration script simple, even if you need to run several “sub-migrations”
```
Comment (ealain) [inline blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md]: Pas actionnable

File: blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+  - Writing integration tests can help catch regressions
```
Comment (ealain) [inline blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md]: Regressions? Not sure what you mean. Is it e2e you are talking about?

File: blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+  - Use dedicated environments to test the migration plan
```
Comment (ealain) [inline blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md]: ...with production data?

File: blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+
+### Backup your data and practice restoring it
+
+If anything goes wrong, you should be *able to restore* your database to a correct state.
+
+- Cloud providers have services dedicated to backups (e.g. [AWS Backup](https://docs.aws.amazon.com/aws-backup/latest/devguide/whatisbackup.html) if you are using AWS, [Actifio](https://www.actifio.com/) on GCP, etc.)
+- Database systems often come with their own backup solution:
+  - PostgreSQL has [`pg_dump`](https://www.postgresql.org/docs/current/backup-dump.html) to export data that can be restored later.
+  - DynamoDB has a specific [backup feature](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/CreateBackup.html)
+  - etc.
```
Comment (ealain) [inline blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md]: Pas utile

File: blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+- Practice restoring your data
+- Be pragmatic. Can you afford to restore your data?
+  - Some data items might have been updated during the process
+  - Is it worth to rollback, if only a few items have an issue?
+  - What is the risk of a data loss?
+
+### Write a “down” migration, as well as the “up” migration… or at least a Plan B
+
+How many times I've heard (or said) "It won't fail, no need of plan B"... and it finally failed ?
+I won't go through implementation details here (and a lot of ORMs have dedicated tools for this), the main point is "Don't forget Murphy's Law!"
```
Comment (ealain) [inline blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md]: C'est vraiment la loi de Murphy? Je mettrais plus le biais cognitif de la pensée au-dessus.

File: blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+
+## 🧱 The plan includes a check that everything is ok in the end
+
+### Keep track of migration states with versioning
+
+- If possible, save the changes and the version of items during the migration process.
+- Keep track of migration applications in each environment.
+- The versions must be strictly increasing (in a specific order relationship) and deterministic to be able to compare versions.
```
Comment (ealain) [inline blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md]: Add an example?

File: blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+Although, this does not guarantee that you will end up with a successful migration after the removal operation, this allows you to detect potential issues you have not forecast and facilitate reverting the changes.
+
+✍️ E.g. only remove ‘address’ after you have correctly added ‘street_name’, ‘city’, ‘postcode’.
+
+### Check that everything is ok **after** the migration
+
+- **Audit the database after migration**
+  - Use a monitoring system to keep track of new errors.
+- Not seeing code or monitoring issue do not mean there is no issue ! Check it with **business owners**.
+- You can set-up end-to-end tests to avoid regressions.
```
Comment (ealain) [inline blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md:178]: Ah found it at the end :)

File: blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+
+## 🌟 What to remember?
+
+These 6 guidelines are just an attempt to sum-up what to care about when applying data migrations. They can also apply to application deployments or whatever operation that introduces a breaking change. But the main learning could also be **capitalize on knowledge** to avoid reproducing the same mistakes !
+
+Want to share your own tips or tech convictions? Don't hesitate to comment on this posts 😉
```
Comment (ealain) [inline blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md]:
```suggestion
Want to share your own tips or tech convictions? Don't hesitate to comment on this post 😉
```

File: blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md (hunk 2 above)
Comment (guiyom-e) [reply blog-posts/six-guidelines-for-risk-less-data-migrations/six-guidelines-for-risk-less-data-migrations.md]: Citation reconstituée de Jonathan Beurel, EM à Theodo :D
